File: projet_groupe4/src/main/java/projet_groupe4/model/Jeu.java
package projet_groupe4.model;

import java.util.List;
import java.util.Set;

import com.fasterxml.jackson.annotation.JsonView;

import jakarta.persistence.CollectionTable;
import jakarta.persistence.Column;
import jakarta.persistence.ElementCollection;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.Table;
import projet_groupe4.view.Views;
@Entity
@Table(name="jeu")
public class Jeu {
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	@JsonView(Views.Common.class)
	private Integer id;
	@Column(name="nom_jeu", columnDefinition="varchar(50)", nullable=false)
	@JsonView(Views.Common.class)
	private String nom;

	@ElementCollection(targetClass = TypeJeu.class)
	@CollectionTable(name = "jeu_types", joinColumns = @JoinColumn(name = "jeu_id"))
	@Enumerated(EnumType.STRING)
	@Column(name = "type_jeu", nullable = false)
	@JsonView(Views.Common.class)
	private Set<TypeJeu> typesJeux;

	@Column(name="age_minimum")
	@JsonView(Views.Common.class)
	private int ageMinimum;
	@Column(name="nombre_joueur_minimum")
	@JsonView(Views.Common.class)
	private int nbJoueurMinimum;
	@Column(name="nombre_joueur_maximum")
	@JsonView(Views.Common.class)
	private int nbJoueurMaximum;
	@Column(name="durée")
	@JsonView(Views.Common.class)
	private int duree;
	@Column(name="nombre_exemplaire")
	@JsonView(Views.Common.class)
	private int nbExemplaire;
	@Column(columnDefinition="DECIMAL(2,1) CHECK (note >= 0 AND note <= 5)")
	@JsonView(Views.Common.class)
	private double note;
	
	@ElementCollection(targetClass = CategorieJeu.class)
	@CollectionTable(name = "jeu_categories", joinColumns = @JoinColumn(name = "jeu_id"))
	@Enumerated(EnumType.STRING)
	@Column(name = "categorie", nullable = false) // juste le nom de la colonne
    @JsonView(Views.Common.class)
	private Set<CategorieJeu> categoriesJeux;
	@Column(name="img_url")
	@JsonView(Views.Common.class)
	private String imgURL;
	@Column(name="besoin_game_master")
	@JsonView(Views.Common.class)
	private boolean besoinGameMaster;
	
	// Constructors
	

	public Jeu(Integer id, String nom, Set<TypeJeu> typesJeux, int ageMinimum, int nbJoueurMinimum, int nbJoueurMaximum,
			int duree, int nbExemplaire, double note, Set<CategorieJeu> categoriesJeux, String imgURL,
			boolean besoinGameMaster) {
		this.id = id;
		this.nom = nom;
		this.typesJeux = typesJeux;
		this.ageMinimum = ageMinimum;
		this.nbJoueurMinimum = nbJoueurMinimum;
		this.nbJoueurMaximum = nbJoueurMaximum;
		this.duree = duree;
		this.nbExemplaire = nbExemplaire;
		this.note = note;
		this.categoriesJeux = categoriesJeux;
		this.imgURL = imgURL;
		this.besoinGameMaster = besoinGameMaster;
	}
	public Jeu(String nom, Set<TypeJeu> typesJeux, int ageMinimum, int nbJoueurMinimum, int nbJoueurMaximum, int duree,
			int nbExemplaire, double note, Set<CategorieJeu> categoriesJeux, String imgURL, boolean besoinGameMaster) {
		this.nom = nom;
		this.typesJeux = typesJeux;
		this.ageMinimum = ageMinimum;
		this.nbJoueurMinimum = nbJoueurMinimum;
		this.nbJoueurMaximum = nbJoueurMaximum;
		this.duree = duree;
		this.nbExemplaire = nbExemplaire;
		this.note = note;
		this.categoriesJeux = categoriesJeux;
		this.imgURL = imgURL;
		this.besoinGameMaster = besoinGameMaster;
	}

	
	
	public Jeu() {
        //TODO Auto-generated constructor stub
    }
    // Getters & Setters
	public Integer getId() {
		return id;
	}
	public void setId(Integer id) {
		this.id = id;
	}
	public String getNom() {
		return nom;
	}
	public void setNom(String nom) {
		this.nom = nom;
	}
	public Set<TypeJeu> getTypesJeux() {
		return typesJeux;
	}
	public void setTypesJeux(Set<TypeJeu> typesJeux) {
		this.typesJeux = typesJeux;
	}
	public int getAgeMinimum() {
		return ageMinimum;
	}
	public void setAgeMinimum(int ageMinimum) {
		this.ageMinimum = ageMinimum;
	}
	public int getNbJoueurMinimum() {
		return nbJoueurMinimum;
	}
	public void setNbJoueurMinimum(int nbJoueurMinimum) {
		this.nbJoueurMinimum = nbJoueurMinimum;
	}
	public int getNbJoueurMaximum() {
		return nbJoueurMaximum;
	}
	public void setNbJoueurMaximum(int nbJoueurMaximum) {
		this.nbJoueurMaximum = nbJoueurMaximum;
	}
	public int getDuree() {
		return duree;
	}
	public void setDuree(int duree) {
		this.duree = duree;
	}
	public int getNbExemplaire() {
		return nbExemplaire;
	}
	public void setNbExemplaire(int nbExemplaire) {
		this.nbExemplaire = nbExemplaire;
	}
	public double getNote() {
		return note;
	}
	public void setNote(double note) {
		this.note = note;
	}
	public Set<CategorieJeu> getCategoriesJeux() {
		return categoriesJeux;
	}
	public void setCategoriesJeux(Set<CategorieJeu> categoriesJeux) {
		this.categoriesJeux = categoriesJeux;
	}
	public String getImgURL() {
		return imgURL;
	}
	public void setImgURL(String imgURL) {
		this.imgURL = imgURL;
	}
	public boolean isBesoinGameMaster() {
		return besoinGameMaster;
	}
	public void setBesoinGameMaster(boolean besoinGameMaster) {
		this.besoinGameMaster = besoinGameMaster;
	}
	public void retirerExemplaires(int i) {
		// TODO Auto-generated method stub
		this.nbExemplaire-=i;
		
	}
	public void ajouterExemplaires(int i) {
		this.nbExemplaire+=i;
	}
	
	// Méthodes utilitaires 
	public boolean estDisponible(List<Emprunt> empruntsActifs, List<Reservation> reservationsActives) {
	    int enUtilisation = 0;
	    
	    if (empruntsActifs != null) {
	        for (Emprunt e : empruntsActifs) {
	            if (e.getJeu().equals(this) &&
	                (e.getStatutLocation() == StatutLocation.enCours || e.getStatutLocation() == StatutLocation.enRetard)) {
	                enUtilisation++;
	            }
	        }
	    }
	    
	    if (reservationsActives != null) {
	        for (Reservation r : reservationsActives) {
	            if (r.getJeu().equals(this) && r.getStatutReservation() == StatutReservation.confirmée) {
	                enUtilisation++;
	            }
	        }
	    }
	    
	    return enUtilisation < nbExemplaire;
	}
	

	
	// toString
	@Override
	public String toString() {
		return "Jeu [id=" + id + ", nom=" + nom + ", typesJeux=" + typesJeux + ", ageMinimum=" + ageMinimum
				+ ", nbJoueurMinimum=" + nbJoueurMinimum + ", nbJoueurMaximum=" + nbJoueurMaximum + ", duree=" + duree
				+ ", nbExemplaire=" + nbExemplaire + ", note=" + note + ", categoriesJeux=" + categoriesJeux
				+ ", imgURL=" + imgURL + ", besoinGameMaster=" + besoinGameMaster + "]";
	}
	
	

}

File: projet_groupe4_angular/src/app/dto/jeu.ts
import { CategorieJeu } from "./enum/categorie-jeu";
import { TypeJeu } from "./enum/type-jeu";

export class Jeu {
    constructor(
    public _id: number,
    public _nom: string,
    public _typesJeux: TypeJeu[] = [],
    public _ageMinimum: number ,
    public _nbJoueurMinimum: number,
    public _nbJoueurMaximum: number,
    public _duree: number,
    public _nbExemplaire: number ,
    public _note: number,
    public _categoriesJeux: CategorieJeu[] = [],
    public _imgURL?: string,
    public _besoinGameMaster: boolean = false
  ) {}

  public get id(): number {
    return this._id;
  }
  public set id(value: number) {
    this._id = value;
  }

  public get nom(): string {
    return this._nom;
  }
  public set nom(value: string) {
    this._nom = value;
  }

  public get typesJeux(): TypeJeu[] {
    return this._typesJeux;
  }
  public set typesJeux(value: TypeJeu[]) {
    this._typesJeux = value;
  }

  public get ageMinimum(): number {
    return this._ageMinimum;
  }
  public set ageMinimum(value: number) {
    this._ageMinimum = value;
  }

  public get nbJoueurMinimum(): number {
    return this._nbJoueurMinimum;
  }
  public set nbJoueurMinimum(value: number) {
    this._nbJoueurMinimum = value;
  }

  public get nbJoueurMaximum(): number {
    return this._nbJoueurMaximum;
  }
  public set nbJoueurMaximum(value: number) {
    this._nbJoueurMaximum = value;
  }

  public get duree(): number {
    return this._duree;
  }
  public set duree(value: number) {
    this._duree = value;
  }

  public get nbExemplaire(): number {
    return this._nbExemplaire;
  }
  public set nbExemplaire(value: number) {
    this._nbExemplaire = value;
  }

  public get note(): number {
    return this._note;
  }
  public set note(value: number) {
    this._note = value;
  }

  public get categoriesJeux(): CategorieJeu[] {
    return this._categoriesJeux;
  }
  public set categoriesJeux(value: CategorieJeu[]) {
    this._categoriesJeux = value;
  }

  public get imgURL(): string | undefined {
    return this._imgURL;
  }
  public set imgURL(value: string | undefined) {
    this._imgURL = value;
  }

  public get besoinGameMaster(): boolean {
    return this._besoinGameMaster;
  }
  public set besoinGameMaster(value: boolean) {
    this._besoinGameMaster = value;
  }

  // toJson()
  public toJson(): any {
    return {
      nom: this._nom,
      typesJeux: this._typesJeux,
      ageMinimum: this._ageMinimum,
      nbJoueurMinimum: this._nbJoueurMinimum,
      nbJoueurMaximum: this._nbJoueurMaximum,
      duree: this._duree,
      nbExemplaire: this._nbExemplaire,
      note: this._note,
      categoriesJeux: this._categoriesJeux,
      imgURL: this._imgURL || null,
      besoinGameMaster: this._besoinGameMaster
    };
  }

}
